File: ios/Kurir/modules/signup/SignupDelegate.swift
//
//  SignupDelegate.swift
//  Kurir
//
//  Created by Loren on 1/9/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Foundation
import Crazybean
import Mobilex

class SignupDelegate: Delegate<SignupView, SignupViewModel> {
    // Switch back to login
    func onLoginClick() {
        view?.showLogin()
    }
    
    func onRegister(mobile: String, email: String) {
        view?.showSpinner()
        viewModel.signup(mobile: mobile, email: email).observe { [weak self] auth in
            if let this = self {
                this.view?.hideSpinner()
                this.view?.showVerify(user: this.viewModel.user)
            }
        }
    }
    
    override func onStop() {
        viewModel.onRelease()
    }
}

File: android/src/main/java/au/com/crazybean/mobilex/kurir/modules/signup/SignupDelegate.kt
package au.com.crazybean.mobilex.kurir.modules.signup

import android.os.Bundle
import androidx.lifecycle.Observer
import au.com.crazybean.mobilex.kurir.data.model.ERR_EXISTS
import au.com.crazybean.mobilex.kurir.data.model.ERR_NOT_FOUND
import au.com.crazybean.sdk.mvvm.Delegate

class SignupDelegate(view: SignupView?,
                     viewModel: SignupViewModel) : Delegate<SignupView, SignupViewModel>(view, viewModel) {
    override fun onCreate(params: Bundle) {
        super.onCreate(params)
        view?.showSignup()
    }

    fun onSignupClick(mobile: String?, email: String?) {
        when {
            mobile.isNullOrBlank() -> Unit // TODO: Show error
            email.isNullOrBlank() -> Unit // TODO: Show error
            else -> viewModel.signup(mobile, email)
                .observe(this, Observer {
                    when (it?.result) {
                        ERR_NOT_FOUND -> view?.showVerify()
                        ERR_EXISTS -> view?.showExists()
                        else -> view?.showError()
                    }
                })
        }
    }

    fun onLoginClick() {
        view?.showLogin()
        view?.dismiss()
    }

    fun onVerifyClick(passcode: String) {
        if (passcode == "123456") {
            view?.showProfile()
        } else {
            view?.showCodeError()
        }
    }

    fun onComplete(firstName: String?, lastName: String?, password: String?) {
        when {
            firstName.isNullOrBlank() -> Unit
            lastName.isNullOrBlank() -> Unit
            password.isNullOrBlank() -> Unit
            else -> viewModel.register(firstName, lastName, password)
                .observe(this, Observer {
                    // Handle the response.
                    view?.showDashboard()
                    view?.dismiss()
                })
        }
    }
}